In [1]:
kernel.silent(true) // supress all non-explicit output
import ammonite.ops._
import sourcecode.Text.generate
val jarString = "../target/scala-3.0.2/euclid-scala-assembly-0.1.0-SNAPSHOT.jar"
val path = ammonite.ops.pwd / ammonite.ops.RelPath(jarString)
interp.load.cp(path)


import ammonite.ops._

import sourcecode.Text.generate

jarString: String = "../target/scala-3.0.2/euclid-scala-assembly-0.1.0-SNAPSHOT.jar"
path: Path = /Users/michaelmccoy/Projects/euclid-scala/target/scala-3.0.2/euclid-scala-assembly-0.1.0-SNAPSHOT.jar

In [2]:
def makeExponents(order: Int, l: List[Int]): List[(Int, Int)] = {
    l.zipWithIndex.map(t => (t._2, t._1)) :::((order, 1) :: Nil)
}

In [3]:
import com.mbmccoy.euclid.polynomial._
import scala.annotation.tailrec

val prime = Prime.next(65537)
val order = 1
val pf = PrimeField(prime)
println(pf)


com.mbmccoy.euclid.polynomial.PrimeField(65537)


In [4]:
def getOrder(x: pf.Element): Int = {
    var y = x
    var order = 1
    while (y != pf.one) {
        y *= x
        order += 1
    }
    order
}
    
// Print some orders here
var x = pf.one
for (_ <- 0 until 10) {
    val order = getOrder(x)
    println(f"x=${x}, order=${order}")
    x = x + pf.one
}

x=1, order=1
x=2, order=32
x=3, order=65536
x=4, order=16
x=5, order=65536
x=6, order=65536
x=7, order=65536
x=8, order=32
x=9, order=32768
x=10, order=65536


In [5]:
val omegaHat = pf(16384) // 128, order = 32
val omega = omegaHat.pow(4) // order = 8
val omegaOrder = getOrder(omega)
val omegaHatOrder = getOrder(omegaHat)
def printOrder(x: pf.Element) = println(f"${x}: order=${getOrder(x)}")

printOrder(omegaHat)
printOrder(omega)


16384: order=16
65281: order=4


In [6]:

val M = pf(44734)
val K1 = pf(34369)
val K2 = pf(51062)
val MInverse = M.reciprocal

// TODO: verify that this is close enough to a width-1 Rescue hash.
def rescueIteration(x: pf.Element): pf.Element = {
    val y: pf.Element = M * x.pow(3) + K1
    val cubeRoot = y.reciprocal.pow((pf.order.toInt - 2)/3)
    cubeRoot * MInverse + K2
}

def rescueHash(x: pf.Element, rounds: Int): pf.Element = {
    if (rounds <= 0) { 
        x
    } else {
        rescueHash(rescueIteration(x), rounds - 1)
    }
}


In [7]:
val fortyTwo = pf(42)

@tailrec
def winterfellReversed(x: pf.Element, rounds: Int, trace: List[pf.Element] = Nil): List[pf.Element] = {
    if (rounds <= 0) trace
    else {
        winterfellReversed(x*x*x + fortyTwo, rounds - 1, x :: trace)
    }
}

In [24]:
//val secret = pf(54234)
val secret = pf(1111)
val executionTrace = winterfellReversed(secret, 4).reverse
println(executionTrace)


List(1111, 34485, 15064, 25005)


In [25]:

def getSubgroup(x: pf.Element) = {
    lazy val subgroup: LazyList[pf.Element] = x #:: subgroup.map{y => x*y}
    pf.one :: subgroup.takeWhile(_ != pf.one).toList
}
val omegas = getSubgroup(omega)

println(omegas)

List(1, 65281, 65536, 256)


In [26]:
println(pf.getClass)
println(pf.one.getClass)
println((-pf.one).getClass)

class com.mbmccoy.euclid.polynomial.PrimeField
class com.mbmccoy.euclid.polynomial.PrimeField$Element
class com.mbmccoy.euclid.polynomial.PrimeField$Element


In [27]:
import com.mbmccoy.euclid.polynomial.FieldOps.lagrangeInterpolation

val alpha = omegas.reduce((a,b) => (pf.one-a)*(pf.one-b))
val alphaInv = alpha.reciprocal
val rootPoly = Polynomial[pf.Element]((0, -pf.one), (omegaOrder, pf.one))
val interpolant = omegas.zip(executionTrace)
.map((w, y) => {
    val poly = rootPoly / (Polynomial[pf.Element]((0, -w), (1, pf.one)))
    y * poly / poly(w)
}).reduce((x, y) => x + y)

println(interpolant)
omegas.zip(executionTrace).map((w, y) => {
    println((interpolant(w), y))
})

2532 + 29783X + 38324X^2 + 61546X^3
(1111,1111)
(34485,34485)
(15064,15064)
(25005,25005)


In [36]:
println("x | f(x)")
println("--|-----")
getSubgroup(omegaHat).zipWithIndex.map((x, i) => {
    println(f"$$\hat{\omega}^{${i}}$$ | ${interpolant(x)}")
})

x | f(x)
--|-----
$\hat{\omega}^{0}$ | 1111
$\hat{\omega}^{1}$ | 6760
$\hat{\omega}^{2}$ | 59183
$\hat{\omega}^{3}$ | 50423
$\hat{\omega}^{4}$ | 34485
$\hat{\omega}^{5}$ | 38024
$\hat{\omega}^{6}$ | 59117
$\hat{\omega}^{7}$ | 20665
$\hat{\omega}^{8}$ | 15064
$\hat{\omega}^{9}$ | 35863
$\hat{\omega}^{10}$ | 50630
$\hat{\omega}^{11}$ | 39013
$\hat{\omega}^{12}$ | 25005
$\hat{\omega}^{13}$ | 60555
$\hat{\omega}^{14}$ | 37809
$\hat{\omega}^{15}$ | 31101


x | f(x)
--|-----
$\hat{\omega}^{0}$ | 1111
$\hat{\omega}^{1}$ | 6760
$\hat{\omega}^{2}$ | 59183
$\hat{\omega}^{3}$ | 50423
$\hat{\omega}^{4}$ | 34485
$\hat{\omega}^{5}$ | 38024
$\hat{\omega}^{6}$ | 59117
$\hat{\omega}^{7}$ | 20665
$\hat{\omega}^{8}$ | 15064
$\hat{\omega}^{9}$ | 35863
$\hat{\omega}^{10}$ | 50630
$\hat{\omega}^{11}$ | 39013
$\hat{\omega}^{12}$ | 25005
$\hat{\omega}^{13}$ | 60555
$\hat{\omega}^{14}$ | 37809
$\hat{\omega}^{15}$ | 31101